designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const report = useStdin
+  ? eslint.executeOnText(text, filePath)
+  : eslint.executeOnFiles(patterns)
+
+for await (const result of report) {
+  print(result)
+}
+```
+
+</details>
+
+#### § The `getFormatter()` method
+
+This method returns a `Promise<Formatter>`. The `Formatter` type is a function `(results: AsyncIterable<LintResult>) => AsyncIterableIterator<string>`. It receives lint results then outputs the formatted text.
+
+This means the `getFormatter()` method wraps the current formatter to adapt the interface.
+
+<details>
+<summary>A rough sketch of the `getFormatter()` method.</summary>
+
+```js
+class ESLint {
+  async getFormatter(name) {
+    const format = this.cliEngine.getFormatter(name)
+
+    // Return the wrapper.
+    return async function* formatter(resultIterator) {
+      // Collect and sort the results.
+      const results = []
+      for await (const result of resultIterator) {
+        results.push(result)
+      }
+      results.sort(ESLint.compareResultsByFilePath)
+
+      // Make `rulesMeta`.
+      const rules = this.cliEngine.getRules()
+      const rulesMeta = getRulesMeta(rules)
+
+      // Format the results with the formatter of the current spec.
+      yield format(results, { rulesMeta })
+    }
+  }
+}
+```
+
+</details>
+
+<details>
+<summary>Example: Use the formatter.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+const results = eslint.executeOnFiles(patterns)
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
+
+Once the `getFormatter()` method got this change, we can update the specification of custom formatters without breakage in the future to support streaming.
+
+#### § The other methods
+
+The following methods receive an `AsyncIterable<LintResult>` object as the first argument then return an `AsyncIterableIterator<LintResult>` object because those methods are sandwiched between `executeOnFiles()` and formatters.
+
+- `outputFixes()`
+- `getErrorResults()`
+
+<details>
+<summary>Example: Use `outputFixes()` and `getErrorResults()`.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+let results = eslint.executeOnFiles(patterns)
+// Update the files of the results if needed
+// (This must be done before `--quiet`)
+if (process.argv.includes("--fix")) {
+    results = ESLint.outputFixes(results)
+}
+// Filter the results if needed
+if (process.argv.includes("--quiet")) {
+    results = ESLint.getErrorResults(results)
+}
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
```

Comment:
This may replicate what exists today, but I find the names of these to be somewhat problematic, but it may be out of scope of this RFC to change. (Though if they should be changed, the time of creating a new class would probably be the time to do it)

Wouldn't something like `decorateWithFixes()` and `decorateWithErrorResults()` be a better reflection of their intention?

Also: What is the reason that these aren't part of `executeOnFiles()` and triggered by some options sent to that method?

Reply:
Good point. Previously, the results are just an array, we hadn't needed to do those in a stream. So those are just small utilities.

But those are used in a stream now. The current API may not fit.